releaser/.dagger/main.go

```diff
@@ -0,0 +1,117 @@
+// A module that encodes the official release process of the Dagger Engine
+package main
+
+import (
+	"context"
+	"dagger/releaser/internal/dagger"
+	"fmt"
+	"net/url"
+	"strings"
+)
+
+type Releaser struct {
+	ChangeNotesDir *dagger.Directory // +private
+}
+
+func New(
+	// +optional
+	// +defaultPath="/"
+	// +ignore=["*", "!.changes/*.md", "!**/.changes/*.md"]
+	changeNotesDir *dagger.Directory,
+) Releaser {
+	return Releaser{
+		ChangeNotesDir: changeNotesDir,
+	}
+}
+
+// Lookup the change notes file for the given component and version
+func (r Releaser) ChangeNotes(
+	ctx context.Context,
+	// The component to look up change notes for
+	// Example: "sdk/php"
+	component,
+	// The version to look up change notes for
+	version string,
+) (*dagger.File, error) {
+	if version == "" {
+		v, err := dag.Version().LastReleaseVersion(ctx)
+		if err != nil {
+			return nil, err
+		}
+		version = v
+	}
+	return r.ChangeNotesDir.File(fmt.Sprintf("%s/.changes/%s.md", component, version)), nil
+}
+
+// Publish a Github release
+func (r Releaser) GithubRelease(
+	ctx context.Context,
+	// GitHub repository URL
+	repository string,
+	// Tag for the GitHub release
+	// eg. sdk/typescript/v0.14.0
+	tag string,
+	// The target for the release
+	// eg. 🤷‍♂️
+	// FIXME: what's the difference with 'tag'? Who knows, it wasn't documented - SH Nov 2024
```

Comment:
This function will create the tag if it doesn't exist.

In this case, `target` is the commit/branch/tag/ref to use, and the `tag` is the new tag to create.